File: iSHARE.IdentityServer/Delegation/DelegationMaskValidationService.cs
using System.Linq;
using iSHARE.Models;
using iSHARE.Models.DelegationEvidence;
using iSHARE.Models.DelegationMask;

namespace iSHARE.IdentityServer.Delegation
{
    public class DelegationMaskValidationService : IDelegationMaskValidationService
    {
        public ValidationResult Validate(DelegationMask delegationMask)
        {
            var policySets = delegationMask.DelegationRequest.PolicySets;

            for (int i = 0; i < policySets.Count; i++)
            {
                var policies = policySets[i].Policies;

                for (int j = 0; j < policies.Count; j++)
                {
                    if (policies[j].Rules.All(r => r.Effect != PolicyRule.Permit().Effect))
                    {
                        return ValidationResult.Invalid(
                            "delegationRequest.policySets[" + i + "].policies[" + j + "] does not contain a rule with the Effect 'Permit'");
                    }
                }
            }

            return ValidationResult.Valid();
        }
    }
}

File: modern/authorization-registry/src/delegation/delegation-mask-validation.service.ts
import { Injectable } from '@nestjs/common';
import { DelegationMask } from './dto/delegation-mask.dto';

export type ValidationResult = { ok: true } | { ok: false; error: string };

/**
 * Port of iSHARE.IdentityServer.Delegation.DelegationMaskValidationService.
 * The mask is only valid if every policy in every policy set asks for at
 * least one rule with effect="Permit" — otherwise no evidence can be issued.
 */
@Injectable()
export class DelegationMaskValidationService {
  validate(mask: DelegationMask): ValidationResult {
    const sets = mask.delegationRequest.policySets;
    for (let i = 0; i < sets.length; i++) {
      const policies = sets[i].policies;
      for (let j = 0; j < policies.length; j++) {
        if (!policies[j].rules.some((r) => r.effect === 'Permit')) {
          return {
            ok: false,
            error: `delegationRequest.policySets[${i}].policies[${j}] does not contain a rule with the Effect 'Permit'`,
          };
        }
      }
    }
    return { ok: true };
  }
}
